Add a changelog and populate it with draft 1.1.0 changes (#85)
ccao-data/ptaxsim: NEWS.md

# ptaxsim 1.1.0

## Background

This release adds support for tax year 2024. It is intended for use with the
`2024.0.x` database version.

The release also features a significant number of changes to the database schema
and to package functions to accommodate changes that the Clerk and the
Treasurer made to the data they report in 2024. Source data from the Clerk and
the Treasurer forms the core of the PTAXSIM database, so any changes they make
need to be reflected in this package, too.

The Clerk and the Treasurer changed their data reporting this past year because
they both completed a migration from a legacy AS/400 mainframe system to a
new integrated property tax system of record.

The main changes that the Clerk and the Treasurer made in 2024 include:

> - **Changed agency numbers to enable reporting some funds as agencies**.

Author: I would love some help filling out the context here!

Reviewer: This one is tricky to summarize in one bullet.. I'm not sure it makes sense to say they "changed agency numbers" but that is the simplest way to put it. 
Is this description to confusing?
* Changed certain agency numbers that are no longer reported as independent agencies (new funds replace what were previously treated as their own taxing agencies, i.e. CITY OF CHICAGO LIBRARY FUND was previously a recorded as a taxing agency; in 2024, all library-related funds are under the taxing agency CITY OF CHICAGO)

Author: Thanks for clarifying this! I took a stab at a better description in 36192dd7eabcf826ed53510229f3987f09899092.

- **Switched from three-digit to six-digit fund numbers to support new types

>   of funds**. Prior to 2024, fund numbers (`agency_fund.fund_num`)

Reviewer: I worry types of funds can be confused with "fund_type" - I would either delete "to support new types of funds" or change to:

> Switched from three-digit to six-digit fund numbers to add greater level of detail

Author: Good call -- I made this change and added a bit more detail to this bullet point in 0c52c66cfefc918331bdd140a9df80bf3220e0ca.

  consisted of three digits, and funds with the same `fund_num` in different
  agencies always referred to the same type of fund. In 2024, the Clerk changed
  their fund numbers so that they consist of six digits, and they are no longer
  guaranteed to refer to the same type of fund across agencies. The new types

>   of funds that this change supports are primarily bond series.

Reviewer: ```suggestion
  consisted of three digits, and funds with the same `fund_num` in different
  agencies always had the same `fund_name`. In 2024, the Clerk changed
  their fund numbers so that they consist of six digits, and they are no longer
  guaranteed to share the same `fund_name` across agencies. The new types
  of funds that this change supports are primarily bond series.
```

Author: Clarifying question before I accept this change: Were the names of funds with the same number indeed guaranteed to be identical prior to 2024? The line on the main branch that calculates `fund_name` in `agency_fund_info` makes it seem like they are perhaps similar, but not identical:

https://github.com/ccao-data/ptaxsim/blob/eb1e3e8d4c7426656ea38b78d9d502b2f06286d2/data-raw/agency/agency.R#L154

Reviewer: My understanding was that this was done because of inconsistencies with case and spacing across years - I think that's ultimately why the table did not originally include `agency_num` as a key.

Author: Gotcha, thanks! I implemented your suggested change in fc6ecf7992979ec6ad35e58d6794b51daa2aa8e0. I think an additional sentence like this might help clarify the situation even further:

```diff
In 2024, the Clerk changed their fund numbers so that they consist of six digits,
and they are no longer guaranteed to share the same name across agencies.
+ They also added many new six-digit funds that were previously aggregated into
+ larger three-digit funds. For these new six-digit funds, the first three digits of the
+ fund number always matches the three-digit number for the aggregate fund that
+ existed prior to 2024.
```

Does this seem correct (and helpful) to you @kyrasturgill?

Reviewer: Yes this makes sense! I was also thinking, should we clarify that for `fund_num` in years to 2024, we added "000" to make the format consistent?

Author: Good idea, done in 30f01dfb54fc8d718b55afb8c5299888c5fbd8ac!

- **Changed the TIF revenue share calculation so that it is performed at the
  PIN-level rather than the tax-code-level**. This means that TIF increment is
  now calculated based on the difference between the frozen and current taxable
  EAV for each individual PIN in a TIF, rather than the frozen and current

>   taxable EAV of the tax code as a whole.

Reviewer: Thoughts on making this bullet first? I feel like it's the most important change they made.

Author: Agreed, I moved it to the top and added a bit more clarifying detail in 61f5b9694170f0359b42926f818f14177c5a7fab.

- **Removed a number of minor columns from agency reports**. None of these
  columns were used in any core PTAXSIM functions, so we expect that their
  absence will not affect most users.
- **Added a new "authority number" for each agency in agency reports**. So far,
  authority numbers appear to have a 1:1 relationship to agency numbers, though
  we expect this may change in future years.

Read on for a detailed description of the changes we made to the PTAXSIM
database and functions to handle these changes in the source data.

## Breaking changes

- **Added new `agency.agency_*_24` columns to handle changing agency numbers

>   in 2024**. You can use these columns to construct a crosswalk to analyze

Reviewer: ```suggestion
- **Added new `agency.agency_*_24` columns to the `agency_info` table to handle changing agency numbers
  in 2024**. You can use these columns to construct a crosswalk to analyze
```

Author: Nice catch! I fixed this in 7ed99ce15c06ad279e4f4b2d95ffaa64404fc91e, since I'm using the schema prefix in the column name to notate the table name (i.e. I changed `agency.agency_*_24` ➡️ `agency_info.agency_*_24`).

  agencies over time, even if their agency number changed in 2024.
    - The new columns include:
      - `agency.agency_change_24` (boolean, required): Whether the agency's
        number changed in 2024.
      - `agency.agency_num_24` (string, optional): The agency's new number
        starting in 2024. Null if the agency number did not change in 2024.
      - `agency.agency_name_24` (string, optional): The agency's name starting in
        2024. Null if the agency number did not change in 2024.
    - **How this change affects you**: If you maintain code that analyzes
      agencies over time, and you want to update your code to include 2024
      data, you will need to use these new columns to handle changes to
      agency numbers. See this vignette for a demonstration of how to
      use the new columns to construct a crosswalk for agency numbers

>       across years: _TK: Link to vignette_

Author: Waiting until #84 lands to add a link here.

Reviewer: Unless we change the .RMD file name, the link should ultimately be https://ccao-data.github.io/ptaxsim/articles/agencies.html

- **Added a new column `agency_fund.fund_type_num` to handle changing fund
  numbers in 2024**. In 2024, the Clerk changed their fund numbers so that
  they consist of six digits instead of three, and they are no longer

>   guaranteed to refer to the same type of fund across agencies. However, the

Reviewer: ```suggestion
  guaranteed to refer to the same `fund_name` across agencies. However, the
```

Author: This is related to my question above, about whether fund names were actually guaranteed to hold across agencies in 2023.

Author: Fixed in fc6ecf7992979ec6ad35e58d6794b51daa2aa8e0.

  first three digits of a fund number are still guaranteed to refer to a
  consistent fund "type", so we introduced a new column
  `agency_fund.fund_type_num` to represent the notion of a "fund type" that
  is consistent across agencies.
    - **How this change affects you**: If you use `agency_fund.fund_num` to

>       track specific types of funds across agencies, you will need to switch to

Reviewer: ```suggestion
      track specific funds across agencies, you will need to switch to
```

  revenue share calculation.** The `tax_bill()` function also accepts an
  optional `pin_tif_dt` argument that you can use for post-2024 TIF
  counterfactuals.
    - Note that there are now two different TIF lookup functions, each of
      which applies to a different set of years:
      - `lookup_pin_tif()` will only return rows where `year >= 2024`.
      - `lookup_tif()` will only return rows where `year < 2024`.
    - **How this change affects you**: If you maintain any TIF counterfactuals,
      you will need to update your use of the `tax_bill()` function to pass in
      an altered `pin_tif_dt` for any tax years that you analyze beyond 2023.
      For an example of this type of change, see the [Tinkering with TIFs
      vignette](https://ccao-data.github.io/ptaxsim/articles/tifs.html), which
      we have updated to include a TIF counterfactual with data for tax year
      2024.
- **Dropped a few columns that the Clerk has removed from its agency reports**.
    - The dropped columns are:
      - `agency_fund.ptell_reduced_ind` (PTELL-reduced levy indicator)
      - `agency.reduction_type` (PTELL-reduced levy type)

>       - `agency.total_non_cap_ext` (Total non-capped extension for an agency)

Author: Is this list of dropped columns correct and comprehensive?

Reviewer: I think my initial documentation mistakenly identified `total_non_cap_ext` as removed, but that one is still in there. `total_reduced_levy` is dropped.

>       - `agency.total_non_cap_ext` (Total non-capped extension for an agency)

Reviewer: ```suggestion
      - `agency_fund.ptell_reduced_ind` (PTELL-reduced levy indicator)
      - `agency.reduction_type` (PTELL-reduced levy type)
      - `agency.total_reduced_levy` (PTELL reduced levy total)
```

    - **How this change affects you**: We recommend ignoring this column for
      now, because it is currently duplicative of `agency_num` and we don't
      have full confidence in our understanding of the Clerk's plans for this
      column in the future. However, it remains available if you are interested
      in investigating it.
- **Added support for pre-release database versions**. The code now supports
  database versions with pre-release suffixes like `2024.0.0-alpha.1`.
  Databases tagged with pre-release versions are unreleased but available for
  testing.
    - **How this change affects you**: This change will not affect most PTAXSIM
      users. It is only useful for PTAXSIM maintainers and pre-release testers.
- **Updated the [Tinkering with TIFs
  vignette](https://ccao-data.github.io/ptaxsim/articles/tifs.html) to
  demonstrate the correct way to handle the new TIF revenue share calcluation**
  ([#77](https://github.com/ccao-data/ptaxsim/pull/77)).
    - **How this change affects you**: You should read the latest version of the
      vignette if you use PTAXSIM for TIF counterfactuals.
- **Added a new vignette to demonstrate the correct way to handle agency

>   numbers that changed in 2024 (_TK: Link to vignette_)**

Author: Same as above: I'm waiting until #84 lands to add this link.
